File: Week1/김효준/괄호.py
import sys
input = sys.stdin.readline

N = int(input())
for _ in range(N):
    PS = input()
    stack = []
    top = -1
    for s in PS:
        if s == '(':
            stack.append(s)
            top += 1
        elif s == ')':
            if len(stack) > 0:
                if stack[top] == '(':
                    stack.pop()
                    top -= 1
                else:
                    stack.append(s)
            else:
                stack.append(s)
    if len(stack) == 0:
        print('YES')
    else:
        print('NO')

File: Week1/장수희/괄호.js
const fs = require("fs");
const input = fs.readFileSync(0, "utf-8").trim().split("\n");

const N = input[0];
const arr = input.slice(1);

for( let i =0; i<N; i++){
    const str = arr[i];
    let stack = [];
    let isVPS = true;
    
    for ( let j=0; j< str.length; j++){
        const ch = str[j]

        if(ch == '('){
            stack.push(ch);
        }
        else{
            if(!stack.length){
                isVPS = false;
               break;  
            }
            stack.pop();
        }
    }
    if(stack.length){
        isVPS = false;
    }
    if (isVPS) console.log("YES");
    else console.log("NO");
}